### ISSUE TITLE
Incorrect collect suggestion - underlying iterator doesn't meet constraints

### ISSUE THREAD

**Issue by jonathandturner** [rust-lang/rust-clippy#7336]
I tried this code:

```rust
    let iter = parser.map(move |contact| match contact {
        Ok(c) => contact_to_value(c, tag.clone()),
        Err(_) => Value::error(ShellError::labeled_error(
            "Could not parse as .vcf",
            "input cannot be parsed as .vcf",
            tag.clone(),
        )),
    });

    let collected: Vec<_> = iter.collect();

    Ok(collected.into_iter().to_output_stream())
```

I expected to see this happen: clippy to notice that the underlying type for the original iterator is not compatible with the constraints for `to_output_stream()`

Instead, this happened: Clippy suggests replacing the collect and using the underlying iterator. This will not compile as the underlying iterator does not meet the constraints required for `to_output_stream()` which include constraints like Send+Sync, which the underlying iterator does not have.

### Meta

- `cargo clippy -V`: e.g. clippy 0.0.212 (f455e46 2020-06-20)
- `rustc -Vv`:
```
rustc 1.54.0-nightly (ed597e7e1 2021-06-08)
binary: rustc
commit-hash: ed597e7e19d0fe716d9f81b1e840a5abbfd7c28d
commit-date: 2021-06-08
host: x86_64-pc-windows-msvc
release: 1.54.0-nightly
LLVM version: 12.0.1
```

<details><summary>Backtrace</summary>
  <p>
  
  ```
  <backtrace>
  ```
  
  </p>
</details>

**Comment by dingxiangfei2009** [comment]
Similarly, this lint does not consider the lifetime constraint as well. One example is the following.

```rust
let u = vec![1,2,3];
let v: Vec<_> = [0,1,2].iter().zip(&u).map(|(x,y)| x+*y).collect(); // (1)
for (x,z) in v.into_iter().zip(u) { // (2)
}
```
`needless_collect` suggests users to remove the `collect` call in (1) and substitute the original iterator into (2), substituting the `into_iter`. However, `u` is moved in (2) while being borrowed in (1), if we do so.

**Comment by aawsome** [comment]
Same problem here.

Using something like
```
let filtered_vec: Vec<_> = vec.into_iter().filter(|item| func(item) ).collect();
foo(filtered_vec.into_iter());
```
as `foo` needs an `ExactSizedIterator`. However, clippy claims we can remove the `collect()`...
